### PR TITLE
restricting file upload types

### LINKED ISSUE #35: Restricting file upload types
A feature to restrict file upload types. Currently, the plugin can only allow additional file upload types.

Slack: https://awesomemotive.slack.com/archives/C2G5XRC9W/p1631056720207600

References: 
https://www.wpbeginner.com/plugins/how-to-restrict-uploads-to-specified-file-types-only-in-wordpress/
https://wordpress.org/plugins/restrict-uploads/

### PR DESCRIPTION
**Pull request by kkarpieszuk**
Fixes #35 

- adds file types natively supported by WP to the Settings > File upload types table so user can now disable them if needed

Test steps:
- activate fut plugin
- go to wp-admin > settings > file upload types and check if file types with description "WordPress natively registered type" are added and automatically enabled
- disable some poplar native file type (for example .jpeg) and save
- install wpforms and create form with file upload field
- preview form and try to upload .jpeg file -> you should see error file extension is not supported
- go to  wp-admin > settings > file upload types and re-enable .jpeg file extension
- go to form and try to upload .jpeg file -> you should be able to do this

### CODE REVIEW

**Review by cheh — CHANGES_REQUESTED** [review]
@kkarpieszuk sorry for being silent for so long time 🙏 and thanks for handling it by yourself!

I've made a first round of reviews and here is my feedback:

1. There is an issue once I've checked out to PR branch without additional actions - *.jpg file type is not allowed, [see](https://a.supportally.com/TFeHTF). The same with other natively registered types, e.g. https://a.supportally.com/4NWVC2, https://a.supportally.com/TUH50E. Env.: WP 5.2 / PHP 5.6.
2. See comment below.

**Comment by kkarpieszuk** [comment]
@cheh I have fixed issue with missing option to upload native type even though it was checked as available

additionally, I have applied all your suggested changes. In a few comments you asked why I decided for some approach and my general answer is: it was because I changed strategy during the work on this and sometimes I forgot to clean the older apporach (like injecting Plugin class into Allowed). I cleared all such cases and now code should be simpler.

**Comment by cheh** [comment]
Unfortunately, there is `PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /plugins/file-upload-types/src/functions.php on line 31`. TBH, I don't have any idea yet why it happens.
One thing which I need to mention - I've caught in on PHP 5.6 / WP 5.2.

**Comment by kkarpieszuk** [comment]
I moved it to separate line for better clarity (also in one more place where I wrapped it in brackets in the same way). @cheh please continue

File: src/Migrations/Dispatcher.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	 * @return void
+	 */
+	public function run_migrations() {
+
+		$already_run    = get_option( 'file_upload_types_migrations_done', [] );
+		$all_migrations = $this->get_migrations_list();
+		$option_changed = false;
+
+		foreach ( $all_migrations as $name => $callback ) {
+			if ( ! isset( $already_run[ $name ] ) && is_callable( $callback ) && $callback() ) {
```
Comment (cheh) [inline src/Migrations/Dispatcher.php:38]: Please, add a blank line before `if`.

File: src/Migrations/Dispatcher.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+					$already_run[ $name ] = 1;
+					$option_changed       = true;
```
Comment (cheh) [inline src/Migrations/Dispatcher.php:40]: Please, remove one extra tab-indent here and above.

File: src/Migrations/Dispatcher.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	 *
+	 * @return void
+	 */
+	public function run_migrations() {
+
+		$already_run    = get_option( 'file_upload_types_migrations_done', [] );
+		$all_migrations = $this->get_migrations_list();
+		$option_changed = false;
+
+		foreach ( $all_migrations as $name => $callback ) {
```
Comment (cheh) [inline src/Migrations/Dispatcher.php:37]: `$this->get_migrations_list()` could be used instead of `$all_migrations`.

File: src/Settings.php [hunk 4]
```diff
@@ -290,6 +292,10 @@ public function display_types_table() { // phpcs:ignore Generic.Metrics.NestingL
 					}
 
 				endif;
+
+					// phpcs:disable WPForms.Comments.PHPDocHooks.RequiredHookDocumentation
+					do_action( 'fileuploadtypes_settings_display_types_table_after_enabled_types' );
```
Comment (cheh) [inline src/Settings.php:297]: `awesomemotive/wpforms-phpcs` package has an issue and notify about an invalid hook name incorrectly. For better readability the `FileUploadTypes` namespace should be converted to `file_upload_types`. 

Could you please change nabes for all new hooks?
Obviously, you will have to use `phpcs:disable` / `phpcs:enable` for reducing a noise.

File: src/Allowed.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+class Allowed {
+
+	/**
+	 * Enabled types.
+	 *
+	 * @since {VERSION}
+	 *
+	 * @var array
+	 */
+	private $enabled_types;
```
Comment (cheh) [inline src/Allowed.php:19]: It could be initialized as an empty array.

**Review by cheh — CHANGES_REQUESTED** [review]
There is an interesting case when the user adds a fake MIME Type with an extension that already exists - https://a.supportally.com/iTIwIg. Personally, I expected that I wasn't able to upload `*.numbers` file.

File: src/Restrict/Native/Admin.php [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+	 * @return void
+	 */
+	public function hooks() {
+
+		add_action( 'fileuploadtypes_settings_display_types_table_after_enabled_types', [ $this, 'table_rows_with_native_types' ] );
+		add_filter( 'fileuploadtypes_migrations_dispatcher_add_native_file_upload_types_callback', [ $this, 'dispatcher_get_migrations_list' ] );
+		add_filter( 'fileuploadtypes_settings_transform_native_type', [ $this, 'get_upload_type' ] );
+	}
+
+	// phpcs:disable Generic.Metrics.CyclomaticComplexity.TooHigh
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply src/Restrict/Native/Admin.php:59]: it is needed in fact: without this, phpcs says the complexity of this method is too high.

File: src/Restrict/Native/Admin.php [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+	 *
+	 * @return string[]
+	 */
+	public function get_types() {
+
+		if ( ! self::$unfiltered_types ) {
+
+			remove_filter( 'upload_mimes', [ $this->allowed, 'allowed_types' ] );
+
+			self::$unfiltered_types = get_allowed_mime_types();
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply src/Restrict/Native/Admin.php:205]: "native" name is a bit misinforming here, but I didn't find anything better. In fact the logic here is to add support for any file types which are not coming from our json file. 

So it doesn't matter if the file type is natively added by WP plugin or some plugin, other than File Upload Types plugin. 

Therefore, the logic here is correct: we just want to get all file types which are not affected by FUT plugin. We should keep this piece of code as is.

File: src/Restrict/Native/Admin.php [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+	public function table_rows_with_native_types() {
+
+		$stored_types  = get_option( 'file_upload_types', [] );
+		$native_types  = isset( $stored_types['native'] ) ? (array) $stored_types['native'] : [];
+		$enabled_types = isset( $stored_types['enabled'] ) ? (array) $stored_types['enabled'] : [];
+
+		if ( ! empty( $native_types ) ) :
+
+		?>
+		<tr class="section">
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply src/Restrict/Native/Admin.php:74]: I can completely do this, but I vote to create an GH issue for this and handle separately as this is not 100% related to the task. I mean this piece of code is, but what do you think about introducing templates for all html we already have? This code to echo pieces of `<table>` is here and there: it would be great to convert them to templates as a separate task, what to you say?

File: src/Migrations/Dispatcher.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ * @since {VERSION}
+ */
+class Dispatcher {
+
+	/**
+	 * Register hooks.
+	 *
+	 * @since {VERSION}
+	 *
+	 * @return void
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply src/Migrations/Dispatcher.php]: changed

File: src/Migrations/Dispatcher.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	/**
+	 * Run all migration logics.
+	 *
+	 * @since {VERSION}
+	 *
+	 * @return void
+	 */
+	public function run_migrations() {
```
[the remark this answers is not in the document]
Comment (kkarpieszuk) [reply src/Migrations/Dispatcher.php]: changed

File: src/Settings.php (hunk 4 above)
Comment (kkarpieszuk) [reply src/Settings.php]: done

**Comment by kkarpieszuk** [comment]
@cheh I've applied most of your suggestions and if I didn't I provided my reasoning comments why not.

about this:

> There is an interesting case when the user adds a fake MIME Type with an extension that already exists - https://a.supportally.com/iTIwIg. Personally, I expected that I wasn't able to upload *.numbers file.

Yup, but the same is in `develop` branch so it is not related to the task. 

In general I found - but I didn't explore it deeply yet, I am still quite new to FUT plugin and mime types handling in WP - WordPress probably in fact ignores mime types when restricting and looks only on file extension. Weird and I might be wrong, but it requires deeper  investigation and that's why I suggest a separate ticket.

So to sum up again:
- I've applied suggested changes when neccessary
- I've provided comments why I can't change some parts
- I suggest two separate tickets: one for html templates and one for the mime issue
